### ISSUE TITLE
Scrollbars should be always visible and instantiated by default on web and desktop

### ISSUE THREAD

**Issue by theacodes** [flutter/flutter#40107]
As mentioned in #38795, the the browser's scrollbar isn't currently visible when content can be scrolled. Ideally, we'd show the scrollbar when content can be scrolled otherwise this breaks users expectations on the platform.

**Comment by sudip4036** [comment]
Any active development is in progress for this feature. We are using Flutter widget as a component in existing web page but the issue is it is disabling browser scrollbar.

**Comment by iapicca** [comment]
HI all,
just my personal opinion, 
if / when a scrollbar for web will be implemented 
I'd love it to have an option to disable it 
thank you

**Comment by Sun3** [comment]
I am trying to get a clarification since there is no milestone setup for the Flutter for the web to have the default scrollbar visible. Is this something that is planned for Flutter for the web to handle?

Thank you.

**Comment by jaredbaszler** [comment]
> I am trying to get a clarification since there is no milestone setup for the Flutter for the web to have the default scrollbar visible. Is this something that is planned for Flutter for the web to handle?
> 
> Thank you.

You'll just have to follow this issue to keep track of the progress.  Since there is no milestone assigned it likely isn't at the top of the list and/or it isn't easy to implement.  So I'd say at this point, I wouldn't expect this to get resolved anytime soon.

**Comment by Sun3** [comment]
@jaredbaszler Thanks...

**Comment by adrian-moisa** [comment]
That's a big issue to address. I picked up flutter with the goal of rendering "web first" and then other devices. Meaning that if web does not work for me it's a big issue. My MVP depends on web working successfully in the early MVP life time and less so on the mobile platforms. Scrolls are super important for UX. I find this is an essential feature to have for the web. Please add priority to this request.

**Comment by amirzolal** [comment]
For the web (and desktop applications), the scrollbars are very important.

**Comment by Piinks** [comment]
Hey @theacodes It looks like the Scrollbar.isAlwaysShown feature reached stable in [version 1.17.](https://github.com/flutter/flutter/commit/ac3b77bdaca677bbea58be11d22317879d363813) Does this feature meet the need you described above? I have tried this on Mac, Web, and mobile, and I think it achieves the effect you described.

```dart
class _MyHomePageState extends State<MyHomePage> {
  final ScrollController scrollController = ScrollController();
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Scrollbar(
        isAlwaysShown: true,
        controller: scrollController,
        child: ListView.builder(
          controller: scrollController,
          itemCount: 200,
          itemBuilder: (context, index) => ListTile(title: Text("Item= ${index + 1}"),),),
      )
    );
  }
}
```

`flutter doctor -v`
``` 
[✓] Flutter (Channel master, 1.24.0-2.0.pre.111, on Mac OS X 10.15.7 19H2 darwin-x64, locale en-US)
    • Flutter version 1.24.0-2.0.pre.111 at /Users/katelovett/github/flutter
    • Framework revision fe189baf34 (14 hours ago), 2020-10-21 22:25:53 -0400
    • Engine revision f459a86610
    • Dart version 2.11.0 (build 2.11.0-242.0.dev)

[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/katelovett/Library/Android/sdk
    ✗ Flutter requires Android SDK 29 and the Android BuildTools 28.0.3
      To update the Android SDK visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions.
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C505
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.1

[✓] Connected device (3 available)
    • macOS (desktop)  • macos      • darwin-x64     • Mac OS X 10.15.7 19H2 darwin-x64
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome 86.0.4240.111

! Doctor found issues in 1 category.
```

**Comment by rbozan** [comment]
> Hey @theacodes It looks like the Scrollbar.isAlwaysShown feature reached stable in [version 1.17.](https://github.com/flutter/flutter/commit/ac3b77bdaca677bbea58be11d22317879d363813) Does this feature meet the need you described above? I have tried this on Mac, Web, and mobile, and I think it achieves the effect you described.
> 
> ```dart
> class _MyHomePageState extends State<MyHomePage> {
>   final ScrollController scrollController = ScrollController();
>   @override
>   Widget build(BuildContext context) {
>     return Scaffold(
>       appBar: AppBar(
>         title: Text(widget.title),
>       ),
>       body: Scrollbar(
>         isAlwaysShown: true,
>         controller: scrollController,
>         child: ListView.builder(
>           controller: scrollController,
>           itemCount: 200,
>           itemBuilder: (context, index) => ListTile(title: Text("Item= ${index + 1}"),),),
>       )
>     );
>   }
> }
> ```
> 
> `flutter doctor -v`
> 
> ```
> [✓] Flutter (Channel master, 1.24.0-2.0.pre.111, on Mac OS X 10.15.7 19H2 darwin-x64, locale en-US)
>     • Flutter version 1.24.0-2.0.pre.111 at /Users/katelovett/github/flutter
>     • Framework revision fe189baf34 (14 hours ago), 2020-10-21 22:25:53 -0400
>     • Engine revision f459a86610
>     • Dart version 2.11.0 (build 2.11.0-242.0.dev)
> 
> [!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
>     • Android SDK at /Users/katelovett/Library/Android/sdk
>     ✗ Flutter requires Android SDK 29 and the Android BuildTools 28.0.3
>       To update the Android SDK visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions.
>     • All Android licenses accepted.
> 
> [✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
>     • Xcode at /Applications/Xcode.app/Contents/Developer
>     • Xcode 11.3.1, Build version 11C505
>     • CocoaPods version 1.9.1
> 
> [✓] Chrome - develop for the web
>     • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome
> 
> [✓] Android Studio (version 4.0)
>     • Android Studio at /Applications/Android Studio.app/Contents
>     • Flutter plugin version 48.1.2
>     • Dart plugin version 193.7547
>     • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
> 
> [✓] VS Code (version 1.49.2)
>     • VS Code at /Applications/Visual Studio Code.app/Contents
>     • Flutter extension version 3.15.1
> 
> [✓] Connected device (3 available)
>     • macOS (desktop)  • macos      • darwin-x64     • Mac OS X 10.15.7 19H2 darwin-x64
>     • Web Server (web) • web-server • web-javascript • Flutter Tools
>     • Chrome (web)     • chrome     • web-javascript • Google Chrome 86.0.4240.111
> 
> ! Doctor found issues in 1 category.
> ```

I am not the mentioned user, but I would like to have the normal scrollbar on web. Things like iOS Safari not scrolling to the top when you press the status bar, or accessibility issues like middle mouse button not functioning (haven’t tried it but I assume it does not work), also the scrollbar arrows where you can press a few times to scroll down and a setting in browsers to have smooth scrolling enabled or disabled are all the things that don’t work with isPersistent.

**Comment by Piinks** [comment]
Hi @rbozan, this issue is focused on scrollbar visibility. 
For interactive scrollbar support, see https://github.com/flutter/flutter/issues/31753
For middle mouse button support, you can follow https://github.com/flutter/flutter/issues/66537
If there are feature requests that you would like to see that aren't covered by an already open issue, feel free to file one so we can start working on finding a solution. Thanks! :)

**Comment by Piinks** [comment]
Another clarifying question - is this a request for always visible scrollbars by default for web and desktop applications? Like how we create the GlowingOverscrollIndicator by default in scroll views?
I have been triaging scrollbar issues this morning and I get the sense this is the full feature expectation.
Also related to: https://github.com/flutter/flutter/issues/36412

**Comment by jlubeck** [comment]
Yes! I think on web and desktop they should be visible by default (when needed). Of course with option to hide

**Comment by theRealBitcoinClub** [comment]
I am also missing the scrollbar as a default for web

**Comment by Sun3** [comment]
For the web version the scrollbar should always be visible if the content does not fit in the current window/page in the browser.

I am really looking forward to this features, since users will not know how to scroll down on a web page when the scrollbars are not visible.

Thank you.

**Comment by Piinks** [comment]
Consolidating into https://github.com/flutter/flutter/issues/70866

**Comment by Piinks** [comment]
Re-opening again as an individual issue

**Comment by Macacoazul01** [comment]
it will be nice to have a new documentation showing how to use the reworked scrollbar on flutter web/desktop after the update launches

**Comment by ganeshrvel** [comment]
The app fails while trying to drag the Scrollbar when used with `SliverList`.

Code:
```dart
  Scrollbar(
      controller: _scrollController,
      child: CustomScrollView(
        controller: _scrollController,
        physics: const ScrollPhysics(),
        slivers: <Widget>[
          SliverList(
            delegate: SliverChildBuilderDelegate((BuildContext context, int index) {
              if (index >= itemLength) {
                return null;
              }

              return Container( width: 20);
            }),
          ),
        ],
      ),
    );
```

If I keep dragging the scrollbar even after reaching the end of the list, it throws the below error: 
```
flutter: [38;5;196m│ [0m[39m[48;5;196m'package:flutter/src/rendering/viewport.dart': Failed assertion: line 1522 pos 12: 'correctedOffset.isFinite': is not true.[0m[49m
flutter: [38;5;196m├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄[0m
flutter: [38;5;196m│ #0   _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)
flutter: [38;5;196m│ #1   _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
flutter: [38;5;196m│ #2   RenderViewport._attemptLayout (package:flutter/src/rendering/viewport.dart:1522:12)
flutter: [38;5;196m│ #3   RenderViewport.performLayout (package:flutter/src/rendering/viewport.dart:1479:20)
flutter: [38;5;196m│ #4   RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1634:7)
flutter: [38;5;196m│ #5   PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:884:18)
flutter: [38;5;196m│ #6   RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:453:19)
flutter: [38;5;196m│ #7   WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:870:13)
flutter: [38;5;196m

```

flutter --version
```
Flutter 1.26.0-2.0.pre.346 • channel master •
https://github.com/flutter/flutter.git
Framework • revision 68d8c4b3bb (5 days ago) • 2021-01-14 12:16:32 -0800
Engine • revision 1b5e16cda2
Tools • Dart 2.12.0 (build 2.12.0-223.0.dev)
```

**Comment by Piinks** [comment]
@ganeshrvel can you please file a new issue with a reproduction of your issue? This issue for for a separate feature. Thank you!

**Comment by ganeshrvel** [comment]
@Piinks created: https://github.com/flutter/flutter/issues/74404

**Comment by arualana** [comment]
> Yes! I think on web and desktop they should be visible by default (when needed). Of course with option to hide

If it when needed, then it is not always visible

**Comment by Leffe108** [comment]
On some platforms (Windows) there is a system preference for if scrollbars should auto-hide or not.

I understand we are not having scrollbars at all and respecting settings like this one is the cherry on top. But just pointing out that this system setting exist. Similar to how an app preferable respect system dark mode setting.

![bild](https://user-images.githubusercontent.com/4630670/105635869-a18cd000-5e65-11eb-95fd-09af17e2c925.png)

**Comment by Piinks** [comment]
I have a proposed solution in #75354 
Although I am not sure this is the best idea to implement so I'm requesting any feedback or input. :)

Here are my concerns with auto-generated Scrollbars:

- users will be annoyed by Scrollbars showing up where they are not wanted
- users that are already using Scrollbars on desktop/web will be broken by duped Scrollbars (if the user implements a Scrollbar, ti will be duped by the auto-scrollbar)
- users implementing or extending ScrollBehavior.buildViewportChrome will be broken
- creating scrollbars on mobile will require the user to add special handling when their app in on web/desktop, lest be affected by the duped scrollbar exception mentioned above
  - or they could turn off autoScrollbars all together and have no issue - which I think kind of nullifies the point of this feature.

Also, I don't think we can automatically detect a Scrollbar created by the user. They can be placed anywhere in the tree, and even if we could track them all down, we would need to distinguish which were made automatically and which were user generated.
Also, there is already a pattern of use where the Scrollbar listens to more than one Scrollable. I've seen this most often with 2D scrolling.

Please see the change (#75354) for more context. All input is welcome while we try to decide the best way to proceed here. :)

**Comment by Leffe108** [comment]
In response to request for comments.

Your [values](https://github.com/flutter/flutter/wiki/Values) state:
> Our first priority is to our developer's user, the end user. Our second priority is to our user, the developer.

On desktop and desktop web end users of software use a varity of input devices. A mouse with a scroll wheel is common, but it cannot be expected that every user has that. For example I use a trackball which do not have a scroll wheel, instead I use the scroll bars and hold a button and scroll by rolling the ball which has infinite movement by design. Other users may be using other devices for accessibility like voice control with eye mouse or a grid mouse. 

Basically, there are a vast vairity of users out there and not every developer is able to understand or anticipate the impact of not adding a scroll bar to their app. So the framework need to be accessible by default in order for most apps to have good accessibility. Then of course I'm not against flags to allow a developer to hide scrollbars if their app is a digital signage app or other situation where they know the scrollbar is not desired. But I think developers should need to opt-out of it.

**Comment by Leffe108** [comment]
In response to request for comments.

Your [values](https://github.com/flutter/flutter/wiki/Values) state:
> Our first priority is to our developer's user, the end user. Our second priority is to our user, the developer.

On desktop and desktop web end users of software use a vairity of input devices. A mouse with a scroll wheel is common, but it cannot be expected that every user has that. For example I use a trackball which do not have a scroll wheel, instead I use the scroll bars and hold a button and scroll by rolling the ball which has infinite movement by design. Other users may be using other devices for accessibility like voice control with eye mouse or a grid mouse. 

Basically, there are a vast vairity of users out there and not every developer is able to understand or anticipate the impact of not adding a scroll bar to their app. So the framework need to be accessible by default in order for most apps to have good accessibility. Then of course I'm not against flags to allow a developer to hide scrollbars if their app is a digital signage app or other situation where they know the scrollbar is not desired. But I think developers should need to opt-out of it. Which could be as simple as a property with a default value.

**Comment by Macacoazul01** [comment]
@Piinks is there any expected date to your pr being implemented on any of the branchs?

**Comment by Sun3** [comment]
I am also looking for an update, especially for Flutter Web. I have a production app that really needs this feature.

Thank you.
